### PR TITLE
Changed signature of operation getConfigStatus for ConfigStatusProvid…

### PR DESCRIPTION
**Pull request by tomhoefer**
…ers so that no i18n handling is required any longer in the binding logic or more precise in the logic of the entity whose config status is to be provided (as discussed in #1033). So the ConfigStatusProvider delivers only a key that will be internationalized by the framework now.

Furthermore the YahooWeatherHandler is now registered as ConfigStatusProvider to be able to provide the configuration status of Yahoo weather things.

Signed-off-by: Thomas Höfer <t.hoefer@telekom.de>

### CODE REVIEW

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java [hunk 1]
```diff
@@ -127,21 +142,66 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        return getThing().getUID().getAsString().equals(entityId);
+    }
+
+    @Override
+    public void setConfigStatusCallback(ConfigStatusCallback configStatusCallback) {
+        this.configStatusCallback = configStatusCallback;
+    }
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters)
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Does this method really have to be implemented? It does not look as if there is anything specific about the Yahoo binding in it - could it be possibly moved to the base implementation or to the framework?

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java [hunk 2]
```diff
@@ -127,21 +142,66 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        return configStatus;
+    }
+
+    @Override
+    public boolean supportsEntity(String entityId) {
+        return getThing().getUID().getAsString().equals(entityId);
+    }
+
+    @Override
+    public void setConfigStatusCallback(ConfigStatusCallback configStatusCallback) {
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Assuming `handleConfigurationUpdate` does not have to be implemented here, this method (together with its field) could be removed as well from the concrete handler implementation.

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java [hunk 3]
```diff
@@ -127,21 +142,66 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+        } catch (IOException e) {
+            logger.debug("Communication error occurred while getting Yahoo weather information.", e);
+        }
+
+        return configStatus;
+    }
+
+    @Override
+    public boolean supportsEntity(String entityId) {
+        return getThing().getUID().getAsString().equals(entityId);
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Same here - having to implement this feels awkward. For the developer, it would be much nicer only having to implement `getConfigStatus` and nothing more.

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java (hunk 1 above)
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Idea: Maybe we could introduce a new `ConfigValidatingThingHandler`which extends `BaseThingHandler`and which implements `ConfigStatusProvider` and with it the three methods that are not specific to a concrete handler implementation? With it, the reference to `ConfigValidationException` could possibly be moved to the new abstract class, so that the `BaseThingHandler` does not introduce any indirect dependency on the `config.validation` package.

File: distribution/smarthome/conf/things/demo.things [hunk 4]
```diff
@@ -1,3 +1,3 @@
-yahooweather:weather:berlin [ location="638242" ]
-yahooweather:weather:losangeles "Los Angeles" [ location="2442047", unit="us", refresh=120 ]
+yahooweather:weather:berlin [ location=638242 ]
+yahooweather:weather:losangeles "Los Angeles" [ location=2442047, unit="us", refresh=120 ]
```
Comment (maggu2810) [inline distribution/smarthome/conf/things/demo.things]: I assume this changes are not related to the config status ones.

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -16,13 +16,13 @@
...
         <config-description>
-            <parameter name="location" type="text" required="true">
+            <parameter name="location" type="integer" required="true">
                 <label>Location</label>
                 <description>Location for the weather information.
                     Syntax is WOEID, see https://en.wikipedia.org/wiki/WOEID.
                 </description>
             </parameter>
-            <parameter name="refresh" type="integer">
+            <parameter name="refresh" type="integer" min="1">
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/ESH-INF/thing/thing-types.xml]: The minimum of the refresh parameter is not related to the config status (or I missed something).

**Comment by maggu2810** [comment]
WRT the few comments, I think the mechanism you implemented is rather clean.

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java (hunk 1 above)
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: :+1:

Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: I assume the current indirect dependency is introduced by "handleConfigurationUpdate":
```java
public void handleConfigurationUpdate(Map<String, Object> configurationParameters)
            throws ConfigValidationException
```

File: distribution/smarthome/conf/things/demo.things (hunk 4 above)
Comment (kaikreuzer) [inline distribution/smarthome/conf/things/demo.things]: They are related to the fact that the Yahoo binding is used as the reference implementation for this feature and as thus it has been cleaned up on the way. Ok for me, but if you prefer, we can also ask @tomhoefer to split it in separate commits.

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java (hunk 1 above)
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Sorry, I am a little bit confused :wink:

File: distribution/smarthome/conf/things/demo.things (hunk 4 above)
Comment (maggu2810) [inline distribution/smarthome/conf/things/demo.things]: Normally I would prefer to use separate commits  for different stuff (so a cleanup should not be mixed with a new feature). The different commits could be (sometime) merged by the same PR, because a PR themselves is not visible in the Git history. It could just confuse people if they are looking at the history of demo things to identify why the quotes has been removed and the commit is "Changed signature of operation getConfigStatus...".

But perhaps we should not be to strict here. :wink:

Comment (tomhoefer) [inline distribution/smarthome/conf/things/demo.things]: Sure I could create a second PR for this purpose. It is up to you ;)

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (tomhoefer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/ESH-INF/thing/thing-types.xml]: exactly... I will remove the changes in the thing type definition and create a separate PR for this purpose

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java (hunk 1 above)
Comment (tomhoefer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: I understand your points... But let us imagine a thing handler is a config status provider and (in future) a FirmwareUpdateHandler as proposed by issue #1153 .... Following your suggestion then we would have to add an additional "FirmwareUpdateThingHandler" implementing the base operations for the FirmwareUpdateHandler.... And if a ThingHandler is both a ConfigStatusProvivder and a FirmwareUpdateHandler what is your suggestion to solve this problem? Shall we offer a base thing handler extension implementing both the ConfigStatusProvider and in future the FirmwareUpdateHandler?

Comment (tomhoefer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: Of course, for now we can start with the mentioned ConfigValidatingThingHandler

Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: I don't think so.
If we would ever use Java 8 we could use default implementations. But I never used this until now and I don't think we should do that.

I recently informed Kai that a binding needs to import the config.validation bundle because the BaseThingHandler throws an exception. We could life with this situation or we could try to solve it. As it is an indirect introduced dependency it is IMHO not a good thing. But I don't know how to solve this cleanly.

File: distribution/smarthome/conf/things/demo.things (hunk 4 above)
Comment (maggu2810) [inline distribution/smarthome/conf/things/demo.things]: Perhaps we can do it after this PR for all the future ones :wink:

Comment (tomhoefer) [inline distribution/smarthome/conf/things/demo.things]: Too late... have already resetted the branch... you will receive a further PR soon :)

File: extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java (hunk 1 above)
Comment (tomhoefer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: You only need to import it when you overwrite the operation... the requirement was that config status and validation also works for other entity types likes rules, not only for things....

Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: > You only need to import it when you overwrite the operation

No, you have to import it in any case if you extend BaseThingHandler, since this introduces an indirect dependency.

Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.yahooweather/src/main/java/org/eclipse/smarthome/binding/yahooweather/handler/YahooWeatherHandler.java]: > But let us imagine a thing handler is a config status provider and (in future) a FirmwareUpdateHandler

I think there will be very very few handlers implementing FirmwareUpdateHandler, so if we need to implement all methods of this interface in that case, so be it.
Mid-term, I agree with @maggu2810 that default implementations of Java 8 might be a good solution to the problem, but we can only go for them once we have increased our minimal requirement.

**Comment by tomhoefer** [comment]
updated this PR

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ConfigStatusThingHandler.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * <li>{@link ConfigStatusProvider#supportsEntity(String)}</li>
+ * <li>{@link ConfigStatusProvider#setConfigStatusCallback(ConfigStatusCallback)}</li>
+ * </ul>
+ * Furthermore it overwrites {@link ThingHandler#handleConfigurationUpdate(Map)} to initiate a propagation of a new
+ * configuration status. So sub classes need only to provide the current configuration status by implementing
+ * {@link ConfigStatusProvider#getConfigStatus()}.
+ *
+ * @author Thomas Höfer - Initial contribution
+ */
+public abstract class ConfigStatusThingHandler extends BaseThingHandler implements ConfigStatusProvider {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ConfigStatusThingHandler.java]: I think `ConfigValidatingThingHandler` would better describe what it is doing.

**Comment by kaikreuzer** [comment]
Thanks. The indirect dependency is not yet addressed, right?

**Comment by tomhoefer** [comment]
I proposed ConfigStatusThingHandler because each BaseThingHandler is also a ConfigValidatingThingHandler... But if you prefer this name then this is also fine for me

**Comment by tomhoefer** [comment]
No, it isnt...

**Comment by kaikreuzer** [comment]
> I proposed ConfigStatusThingHandler

Sorry, I must have missed that. 
@maggu2810 Any preference?

> No, it isnt...

Because
a) you do not agree that should be done
b) you are currently working on it and will update this PR
c) you suggest to do it in a follow-up PR
?

**Comment by tomhoefer** [comment]
To be honest I dont want to change it now because I dont have any idea how to fix this issue. As I already mentioned the requirement was that we should support all entity types that have a configuration. If you like then I will create an issue

**Comment by maggu2810** [comment]
Do we plan that this class should be used by handlers that provide a configuration status or by handlers that could be configured? I do not know how we would like to drop the config validation exception at all. The current implementation in the base thing handler is used by bindings, so I assume we cannot move this function from this class (as sub classes of the BaseThingHandler could rely on this implementation). So ConfigurableThingHandler seems also not be the best one.

**Comment by tomhoefer** [comment]
We plan use this class for thing handlers that provides a config status...

**Comment by tomhoefer** [comment]
I also don´t see any possibility to avoid the dependency on config.validation (of course, for the moment we could move it to the things package, but I think that we want to validate later config parameters of rules as well in our project)

**Comment by kaikreuzer** [comment]
One option would be to not have the exception in the method signature by changing it to a RuntimeException. Usually, I am not a big fan of runtime exceptions, but as this method is ONLY called by the framework, which is under our control, this would be a valid option, since we know that we should expect it to be thrown.

**Comment by tomhoefer** [comment]
Of course, this would solve the problem. But what do you mean by framework in this context? E.g. in our project we call the operation ThingHandler.handleConfigurationUpdate in so-called step handlers of our generic configuration flow and of course we catch the ConfigValidationException. Would you consider this then also as part of the framework? Other callers may not know that such an exception is thrown...

**Comment by kaikreuzer** [comment]
Ok, you are right, I somehow had in mind that the ThingManager calls it, but this is indeed not the case.
So it is some middle ground - no "extension" will ever call it, so it is only proprietary core extensions that might have to deal with that; I would still say that this is ok in that case and if we mention it in the Javadoc, this should suffice.

**Comment by tomhoefer** [comment]
ok... do you want me to change this also in this PR?

**Comment by tomhoefer** [comment]
I would prefer to change this in a follow-up PR

**Comment by kaikreuzer** [comment]
OK for me.
